File: Find_All_Anagrams_in_a_String.py
# Time Complexity : O(n + m), where n is the length of s and m is the length of p
# Space Complexity : O(m), where m is the length of p
from collections import Counter
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        p_count = Counter(p)
        window_count = Counter(s[:len(p)])
        result = []
        
        if p_count == window_count:
            result.append(0)
        
        for i in range(1, len(s) - len(p) + 1):
            if s[i - 1] in window_count:
                window_count[s[i - 1]] -= 1
                if window_count[s[i - 1]] == 0:
                    del window_count[s[i - 1]]
            
            if s[i + len(p) - 1] in window_count:
                window_count[s[i + len(p) - 1]] += 1
            else:
                window_count[s[i + len(p) - 1]] = 1
            
            if p_count == window_count:
                result.append(i)
        
        return result

File: Find All Anagrams in a String.cpp
//Time Complexity- O(s.length+p.length)
//Space Complexity- O(1)

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        
        if(p.size()>s.size()){
            return {};
        }
        
        unordered_map<char,int> mps;
        unordered_map<char,int> mpp;
        
        for(char ch:p){
            mpp[ch]++;
        }
        
        vector<int> ans; 
        for(int i=0;i<p.size();i++){
            mps[s[i]]++;
        }
        
        for(int i=p.size();i<s.size();i++){
            if(mpp==mps){
                ans.push_back(i-p.size());
            }
            mps[s[i-p.size()]]--;
            if(mps[s[i-p.size()]]==0){
                mps.erase(s[i-p.size()]);
            }
            mps[s[i]]++;
        }
        if(mpp==mps){
            ans.push_back(s.size()-p.size());
        }
        return ans;
    }
};